### PR TITLE
F24/trinity/create password backend

### PR DESCRIPTION
**Pull request by trinity-y**
## Notion ticket link

[Set New Password Functionality](https://www.notion.so/uwblueprintexecs/Sprint-Board-4f2ffd2639ac41c4ad0f64818d84d5fa?p=12210f3fb1dc80ed98b2e0fa0e28a6eb&pm=s)

## Implementation description
* If the user is logged in, they can access the create-password-page.
* They can then create a new password. An appropriate error message will pop up if not enough characters are chosen, if the passwords don't match, or if some other unknown error pops up. 
* If the password reset is successful, the user is logged in and a success modal pops up. If the user's status is invited, it is set to active.

The page is viewable on /create-password.

**Testing Private Routes**
1. Log out of your current account by deleting 'authenticateduser' from localstorage.
2. Test that you can't visit the create password page.

**Testing reset password**
1. Log in 
2. Visit create password page and reset password (you can also test mismatching passwords + password length here)
3. To test that the automatic log in was successful, you can refresh the page and try to reset the password again. If you haven't been logged in successfully, it should cause an unknown error.
4. Test that the password reset was successful by logging out (delete authenticated user) and then signing in with the new password.
5. In your DB, check that your user status is now set to ACTIVE.

entering+resetting password
9918bb38-c63d-40d0-9c52-a114d18cd730
(refreshing and trying to reset password again shows that automatic log in was successful) (sorry the gif quality is horrible)

## What should reviewers focus on?
* made the page a private route, rather than redirecting before page load

### CODE REVIEW

**Review by jerry-cheng5 — APPROVED** [review]
Really great work! Everything works on my end. The code looks pretty solid too.

Fingers crossed for an easy merge into main :)

File: backend/typescript/rest/authRoutes.ts [hunk 1]
```diff
@@ -118,4 +118,42 @@ authRouter.post(
   },
 );
 
+authRouter.post(
+  "/setPassword/:email",
+  isAuthorizedByEmail("email"),
```
[the remark this answers is not in the document]
Comment (trinity-y) [reply backend/typescript/rest/authRoutes.ts:123]: just tested, it's unauthorized 👍

File: backend/typescript/rest/authRoutes.ts [hunk 2]
```diff
@@ -118,4 +118,42 @@ authRouter.post(
...
+      );
+      if (responseSuccess.success) {
+        const user = await userService.getUserByEmail(req.params.email);
+        if (user.status === UserStatus.INVITED) {
+          userService.updateUserById(user.id, {
+            ...user,
+            status: UserStatus.ACTIVE,
+          });
+        }
+        // automatically log in after password reset
```
Comment (sthuray) [inline backend/typescript/rest/authRoutes.ts:138]: /auth/setPassword is doing 3 different things at once (setting password, updating userDTO, and logging in the user), imo the latter two actions should be handled as their own calls in the frontend (posting to "/user" and "/auth/login". That way we can handle the individual errors from each of these actions (and less code redundancy).

I think it makes more sense for this endpoint to only return the first responseSuccess from authService.setPassword (line 126).

Comment (trinity-y) [reply backend/typescript/rest/authRoutes.ts]: updated — status update cannot easily be separated from password due to authentication, but I separated login + set password.

**Comment by trinity-y** [comment]
@jerry-cheng5 @sthuray did you want to take a look after the new changes?

File: backend/typescript/services/implementations/authService.ts [hunk 3]
```diff
@@ -222,6 +222,29 @@ class AuthService implements IAuthService {
...
+      const uid = await (await firebaseAdmin.auth().getUserByEmail(email)).uid;
+      await firebaseAdmin.auth().updateUser(uid, {
+        password: newPassword,
+      });
+      return { success: true } as ResponseSuccessDTO;
+    } catch (error: any) {
+      Logger.error(`Failed to update password. Error: ${error}`);
+      if (error.code === "auth/invalid-password") {
+        errorMessage =
+          "Invalid password. Please make sure your password is at least 6 characters!";
```
Comment (MinIcecream) [inline backend/typescript/services/implementations/authService.ts:241]: Wait how come here it's minimum 6 characters but in the frontend we make it minimum 8?

File: backend/typescript/rest/authRoutes.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (jerry-cheng5) [reply backend/typescript/rest/authRoutes.ts:124]: nice nice

**Comment by jerry-cheng5** [comment]
Looks good to me!

File: backend/typescript/rest/authRoutes.ts [hunk 4]
```diff
@@ -118,4 +118,30 @@ authRouter.post(
...
+authRouter.post(
+  "/setPassword/:email",
+  isAuthorizedByEmail("email"),
+  async (req, res) => {
+    try {
+      const responseSuccess = await authService.setPassword(
+        req.params.email,
+        req.body.newPassword,
+      );
+      if (responseSuccess.success) {
```
Comment (sthuray) [inline backend/typescript/rest/authRoutes.ts:131]: Does this handle the case that where this is false?

**Pull request merged**